src/array.jl

```diff
@@ -419,43 +420,49 @@ function copyto!(dest::CatArrOrSub{T, N}, dstart::Integer,
     dpool = pool(dest)
     spool = pool(src)
 
-    # try converting src to dest type to avoid partial copy corruption of dest
-    # in the event that the src cannot be copied into dest
-    slevs = convert(Vector{T}, levels(src))
-    if eltype(src) >: Missing && !(eltype(dest) >: Missing) && !all(x -> x > 0, srefs)
-        throw(MissingException("cannot copy array with missing values to an array with element type $T"))
-    end
-
-    newlevels, ordered = mergelevels(isordered(dest), levels(dest), slevs)
-    # Orderedness cannot be preserved if the source was unordered and new levels
-    # need to be added: new comparisons would only be based on the source's order
-    # (this is consistent with what happens when adding a new level via setindex!)
-    ordered &= isordered(src) | (length(newlevels) == length(levels(dest)))
-    if ordered != isordered(dest)
-        isa(dest, SubArray) && throw(ArgumentError("cannot set ordered=$ordered on dest SubArray as it would affect the parent. Found when trying to set levels to $newlevels."))
-        ordered!(dest, ordered)
-    end
-
-    # Simple case: replace all values
-    if !isa(dest, SubArray) && dstart == dstart == 1 && n == length(dest) == length(src)
-        # Set index to reflect refs
-        levels!(dpool, T[]) # Needed in case src and dest share some levels
-        levels!(dpool, index(spool))
-
-        # Set final levels in their visible order
-        levels!(dpool, newlevels)
-
-        copyto!(drefs, srefs)
-    else # More work to do: preserve some values (and therefore index)
-        levels!(dpool, newlevels)
-
-        indexmap = indexin(index(spool), index(dpool))
-
-        @inbounds for i = 0:(n-1)
-            x = srefs[sstart+i]
-            drefs[dstart+i] = x > 0 ? indexmap[x] : 0
+    if isordered(dest)
+        remap = Vector{eltype(drefs)}(undef, length(levels(src)))
+        seen = falses(length(levels(src))) # needed as we have to dynamically collect levels
+        @inbounds for i in 0:n-1
+            s = srefs[sstart+i]
+            if eltype(src) >: Missing && s == 0
+                if eltype(dest) >: Missing
+                    drefs[dstart+i] = 0
+                else
+                    throw(MissingException("cannot copy array with missing values to an array with element type $T"))
+                end
+            else
+                if seen[s]
+                    drefs[dstart+i] = remap[s]
+                else
+                    seen[s] = true
+                    rm = dpool.invindex[s]
```

Comment:
AFAICT this doesn't work, and better avoid using internal fields:
```suggestion
                    rm = index(dpool)[s]
```
(The call to `index` would better be hoisted out of the loop.)

Reply:
Thanks - this is the part you know best. I will also move it out from the loop.